### DISCUSSION TITLE
How to Restrict Socket Connection to Authenticated Users

### DISCUSSION THREAD

**Discussion by Ujjwal091** [strapi-community/plugin-io#70] [Q&A] [answered]
Hello,

I'm currently working on a chat application and I'm trying to ensure that only authenticated users can connect to the socket. However, I haven't found a direct way to prevent unauthorized connections. Presently, my approach involves allowing connections and then disconnecting unauthorized users, but I'd prefer a method to prevent unauthorized connections altogether.

Could anyone guide me on implementing a setup where only authenticated users can establish socket connections, avoiding the need to disconnect them after they connect?

Thank you for any insights or suggestions you can provide!

**Comment by ComfortablyCoding** [comment]
What is the use case? 

Any connections with the public role will have the same access as whatever the role has currently with the rest api. They wouldn't be receiving any additional data.

**Reply by Ujjwal091** [reply]
The primary objective of restricting socket connections to authenticated users in our chat application is to bolster security and enforce access control. By requiring authentication before establishing a socket connection, we aim to prevent unauthorized access to our real-time chat feature.

Our intention is to ensure that only users who have successfully gone through the authentication process, possibly utilizing JWT tokens or similar mechanisms, can access the chat functionality. This aligns with our broader security measures and follows a principle of least privilege, allowing access only to authenticated and authorized individuals.

Regarding the "public role" and its access level, the concern isn't about providing additional data but rather about the initial access to the socket connection itself. We want to ensure that even users with the public role need to authenticate before utilizing the chat feature, adhering to our security protocols.

Also if there are suggestions or best practices regarding implementing authentication for socket connections within the context of a Flutter application that you could recommend, it would be incredibly helpful.

Thank you for your assistance and insights into this matter.

**Reply by ComfortablyCoding** [reply]
I believe a middleware should be the optimal solution here. I'll see if I can post an example here.

Unfortunately I do not have a any experience in context with flutter at this time so cannot help in that regard.

**Reply by ComfortablyCoding** [reply, marked as answer]
A sample code would be something along the lines of
```js
io: {
  // ...
    config: {
    // ...
      hooks: {
        init({ $io }) {
          $io.server.use((socket,next) => {
            // check if socket connected with valid role.
            const isValidRole = .....;
            if(isValidRole){
              next();
            } else {
              next(new Error('Invalid Auth'));
            }
          });
        }
      }
    // ...  
    }
  // ...
}
```
This middleware will run after the initial handshake takes place. Hope that helps.
